File: client/src/services/commentService.ts
/**
 * Comment API Service — chamadas REST para o backend de comentários.
 *
 * Usa o axios `api` centralizado que já injeta o JWT do Clerk.
 */
import { api } from './api';

// ── Types ─────────────────────────────────────────────────────────────────

export interface CommentCreatePayload {
    anchor_key: string;
    selected_text: string;
    body: string;
    is_private: boolean;
    user_name?: string;
    user_image_url?: string;
}

export interface CommentOut {
    id: number;
    tenant_id: string;
    user_id: string;
    anchor_key: string;
    selected_text: string;
    body: string;
    status: 'pending' | 'approved' | 'rejected' | 'private';
    created_at: string;   // ISO date string from backend
    updated_at: string;
    moderated_by: string | null;
    moderated_at: string | null;
    // Profile data (populated by backend)
    user_name: string | null;
    user_image_url: string | null;
}

// ── API calls ─────────────────────────────────────────────────────────────

/**
 * Cria um comentário no backend.
 */
export async function createComment(payload: CommentCreatePayload): Promise<CommentOut> {
    const { data } = await api.post<CommentOut>('/comments/', payload);
    return data;
}

/**
 * Busca comentários por anchor_key (aprovados + privados do usuário).
 */
export async function fetchCommentsByAnchor(anchorKey: string): Promise<CommentOut[]> {
    const { data } = await api.get<CommentOut[]>(
        `/comments/anchor/${encodeURIComponent(anchorKey)}`
    );
    return data;
}

/**
 * Busca comentários de múltiplos anchors em paralelo.
 * Retorna Map<anchorKey, CommentOut[]>.
 */
export async function fetchCommentsByAnchors(
    anchorKeys: string[],
): Promise<Map<string, CommentOut[]>> {
    const unique = [...new Set(anchorKeys)];
    const results = await Promise.allSettled(
        unique.map(key => fetchCommentsByAnchor(key)),
    );

    const map = new Map<string, CommentOut[]>();
    results.forEach((result, i) => {
        if (result.status === 'fulfilled') {
            map.set(unique[i], result.value);
        } else {
            console.warn(`[commentService] Failed to fetch comments for anchor "${unique[i]}":`, result.reason);
            map.set(unique[i], []);
        }
    });
    return map;
}

// ── Edição / Exclusão (autor) ─────────────────────────────────────────────

/**
 * Atualiza o corpo de um comentário existente (somente autor).
 * Comentários aprovados voltam para "pending" após edição.
 */
export async function updateComment(commentId: number, body: string): Promise<CommentOut> {
    const { data } = await api.patch<CommentOut>(`/comments/${commentId}`, { body });
    return data;
}

/**
 * Remove permanentemente um comentário (somente autor).
 */
export async function deleteComment(commentId: number): Promise<void> {
    await api.delete(`/comments/${commentId}`);
}

// ── Admin / Moderação ─────────────────────────────────────────────────────

/**
 * [Admin] Lista todos os comentários pendentes de moderação.
 */
export async function fetchPendingComments(): Promise<CommentOut[]> {
    const { data } = await api.get<CommentOut[]>('/comments/admin/pending');
    return data;
}

/**
 * [Admin] Aprova ou rejeita um comentário.
 */
export async function moderateComment(
    commentId: number,
    action: 'approve' | 'reject',
    note?: string,
): Promise<CommentOut> {
    const { data } = await api.patch<CommentOut>(
        `/comments/admin/${commentId}`,
        { action, note },
    );
    return data;
}

/**
 * Lista anchor_keys que possuem comentários aprovados.
 * Usado pelo frontend para aplicar a classe .has-comment.
 */
export async function fetchCommentedAnchors(): Promise<string[]> {
    const { data } = await api.get<string[]>('/comments/anchors');
    return data;
}

File: backend/services/comment_service.py
"""
Service de Comentários — orquestra regras de negócio.

Segue o padrão dos outros services do projeto: camada intermediária
entre routers e repositories, responsável por validações e lógica de domínio.
"""

import logging
from sqlalchemy.ext.asyncio import AsyncSession

from backend.domain.comment_models import Comment
from backend.infrastructure.repositories.comment_repository import CommentRepository
from backend.presentation.schemas.comment_schemas import (
    CommentCreate,
    CommentApproveIn,
    CommentUpdate,
)

logger = logging.getLogger("service.comments")


class CommentService:
    def __init__(self, session: AsyncSession):
        self.repo = CommentRepository(session)

    async def create_comment(
        self,
        data: CommentCreate,
        tenant_id: str,
        user_id: str,
    ) -> Comment:
        """
        Cria um comentário.
        - is_private=True  → status 'private'  (nunca vai para moderação)
        - is_private=False → status 'pending'   (aguarda aprovação admin)
        """
        status = "private" if data.is_private else "pending"
        comment = Comment(
            tenant_id=tenant_id,
            user_id=user_id,
            anchor_key=data.anchor_key,
            selected_text=data.selected_text,
            body=data.body,
            status=status,
            user_name=data.user_name,
            user_image_url=data.user_image_url,
        )
        created = await self.repo.create(comment)
        logger.info(
            "Comentário criado: id=%s tenant=%s user=%s anchor=%s status=%s",
            created.id,
            tenant_id,
            user_id,
            data.anchor_key,
            status,
        )
        return created

    async def list_for_anchor(
        self,
        tenant_id: str,
        anchor_key: str,
        user_id: str,
    ) -> list[Comment]:
        """Comentários aprovados + privados do próprio usuário para o anchor."""
        return await self.repo.list_by_anchor(tenant_id, anchor_key, user_id)

    async def list_pending(self, tenant_id: str) -> list[Comment]:
        """Todos os comentários pendentes para o admin moderar."""
        return await self.repo.list_pending(tenant_id)

    async def moderate(
        self,
        comment_id: int,
        data: CommentApproveIn,
        tenant_id: str,
        admin_user_id: str,
    ) -> Comment:
        """
        Aprova ou rejeita um comentário.
        Somente comentários do mesmo tenant podem ser moderados.
        """
        comment = await self.repo.get_by_id(comment_id)
        if not comment:
            raise ValueError(f"Comentário {comment_id} não encontrado")
        if comment.tenant_id != tenant_id:
            raise PermissionError("Sem permissão para moderar este comentário")

        new_status = "approved" if data.action == "approve" else "rejected"
        updated = await self.repo.update_status(
            comment,
            status=new_status,
            moderated_by=admin_user_id,
            note=data.note,
        )
        logger.info("Comentário %s → %s por %s", comment_id, new_status, admin_user_id)
        return updated

    async def update_comment(
        self,
        comment_id: int,
        data: CommentUpdate,
        tenant_id: str,
        user_id: str,
    ) -> Comment:
        """
        Edita o corpo de um comentário.
        Somente o autor pode editar. Se era aprovado, volta para 'pending'.
        """
        comment = await self.repo.get_by_id(comment_id)
        if not comment:
            raise ValueError(f"Comentário {comment_id} não encontrado")
        if comment.tenant_id != tenant_id:
            raise PermissionError("Sem permissão")
        if comment.user_id != user_id:
            raise PermissionError("Somente o autor pode editar")
        if comment.status == "rejected":
            raise ValueError("Comentário rejeitado não pode ser editado")

        updated = await self.repo.update_body(comment, data.body)
        logger.info("Comentário %s editado por %s", comment_id, user_id)
        return updated

    async def delete_comment(
        self,
        comment_id: int,
        tenant_id: str,
        user_id: str,
    ) -> None:
        """
        Remove definitivamente um comentário.
        Somente o autor pode deletar.
        """
        comment = await self.repo.get_by_id(comment_id)
        if not comment:
            raise ValueError(f"Comentário {comment_id} não encontrado")
        if comment.tenant_id != tenant_id:
            raise PermissionError("Sem permissão")
        if comment.user_id != user_id:
            raise PermissionError("Somente o autor pode deletar")

        await self.repo.delete(comment)
        logger.info("Comentário %s deletado por %s", comment_id, user_id)

    async def get_commented_anchors(self, tenant_id: str) -> list[str]:
        """Lista de anchor_keys com comentários aprovados (para o renderer)."""
        return await self.repo.list_anchors_with_comments(tenant_id)
